openshift-kube-apiserver/openshiftkubeapiserver/patch_handlerchain.go

```diff
@@ -95,3 +105,104 @@ func translateLegacyScopeImpersonation(handler http.Handler) http.Handler {
+// 3. no limit.  We have this to handle cases of large clusters with more than 20000 mounted secrets and configmaps.
+//    I honestly don't know how common this is, but I don't want to break on it.
+// Recall that we observed more than the the 30,000 per minute observed during some disruptive events on a cluster.
+// In addition, we special case watches in the platform operator namespaces, cluster scope, and kube-system.
+// We have not observed large numbers of these and they are required in order to make progress when trying to correct
+// some kinds of cluster failures.
+func withWatchRateLimit(handler http.Handler) http.Handler {
+	earlyRateLimiter := flowcontrol.NewTokenBucketRateLimiter(16.6, 100)
+	middleRateLimiter := flowcontrol.NewTokenBucketRateLimiter(33.3, 100)
+
+	// after 20 minutes, we eliminate the rate limiter.  This lets us tolerate very large clusters, while having a ramp
+	// up period during which we can make forward progress
+	startTime := time.Now()
+	earlyEndTime := startTime.Add(10 * time.Minute)
+	middleEndTime := startTime.Add(20 * time.Minute)
+
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		var effectiveRateLimiter flowcontrol.RateLimiter
+		now := time.Now()
+		switch {
+		case now.After(middleEndTime):
+			// we are past our rate limiting time
+			handler.ServeHTTP(w, req)
+			return
+		case now.After(earlyEndTime):
+			effectiveRateLimiter = middleRateLimiter
+		default:
+			effectiveRateLimiter = earlyRateLimiter
+		}
+
+		ctx := req.Context()
+
+		attributes, err := filters.GetAuthorizerAttributes(ctx)
+		if err != nil {
+			// if we cannot get attributes, don't fail the request
+			handler.ServeHTTP(w, req)
+			return
+		}
+		if attributes.GetUser() == nil {
+			// if we cannot get user, don't fail the request
+			handler.ServeHTTP(w, req)
+			return
+		}
+		for _, group := range attributes.GetUser().GetGroups() {
+			if group == user.SystemPrivilegedGroup {
+				// system:masters always have the power!
+				handler.ServeHTTP(w, req)
+				return
+			}
+		}
+
+		if attributes.GetVerb() != "watch" {
+			// only throttle watch establishment
+			handler.ServeHTTP(w, req)
+			return
+		}
+		namespace := attributes.GetNamespace()
+		switch {
+		case len(namespace) == 0:
+			// don't rate limit cluster scoped watches because operators that need to make progress may use those
+			// if we have to restrict further we can
+			handler.ServeHTTP(w, req)
+			return
+
+		case strings.HasPrefix("kube-", namespace):
+			// don't rate limit kube- because some operators use and we use this for delegated authn
+			handler.ServeHTTP(w, req)
+			return
+
+		case strings.HasPrefix("openshift-", namespace):
+			// don't rate limit openshift- because we need operators to make progress, so we need openshift- mounts
+			// to succeed in order to progress
+			handler.ServeHTTP(w, req)
+			return
+
+		}
+
+		// TODO Abu: Can I get some metrics help here?
```

Comment:
right now when p&f rejects a request, we do:
```
metricd.RecordRequestTermination(r, requestInfo, metrics.APIServerComponent, http.StatusTooManyRequests)
```

I am wondering if we should use the same metrics - the only difference is this is a long running request and I don't think we will be able to differentiate whether it is `watch`.  The `verb` will always show `GET` for this metric. 

Maybe we can use a different component name here, `apiserver-watch` so we can make the distinction. I think keeping the same metrics for all rejections make sense, thoughts?